File: server-java/src/main/java/com/aitrack/server/domain/port/EditRecordPort.java
package com.aitrack.server.domain.port;

import com.aitrack.server.domain.model.EditRecordEntity;
import com.aitrack.server.domain.model.PageResult;

import java.time.Instant;
import java.util.List;

/**
 * Driven-side persistence port for edit records.
 *
 * <p>This is a pure domain interface (a secondary port of the hexagon).
 * The {@code adapter.db.EditRecordRepository} Spring Data interface provides the
 * concrete implementation; domain and application code depend only on this port.
 *
 * <p>Pagination is expressed with plain {@code int page} / {@code int size} primitives
 * rather than {@code org.springframework.data.domain.Pageable} to prevent Spring Data
 * infrastructure types from leaking into the domain layer.
 */
public interface EditRecordPort {

    /** Persists a single edit record; returns the saved instance. */
    EditRecordEntity save(EditRecordEntity record);

    /** Counts records for a token+file within a rolling window (rate limiting). */
    long countByTokenKeyAndFilePathSince(String tokenKey, String filePath, Instant since);

    /** Returns a page of records optionally filtered by token key and repo URL. */
    PageResult<EditRecordEntity> findByFilters(String tokenKey, String repoUrl, int page, int size);

    /** Loads all non-rejected records for a token, excluding the given status. */
    List<EditRecordEntity> findByTokenKeyAndStatusNot(String tokenKey, EditRecordEntity.RecordStatus status);

    /** Aggregates stats grouped by token key. */
    List<Object[]> aggregateByTokenKey();

    /** Aggregates stats grouped by repo URL. */
    List<Object[]> aggregateByRepo();

    /** Aggregates stats grouped by device ID. */
    List<Object[]> aggregateByDevice();

    /** Aggregates stats grouped by hostname. */
    List<Object[]> aggregateByHostname();
}

File: server-go/internal/domain/port/edit_record_port.go
// Package port defines the driven-side interfaces (secondary ports) of the
// hexagon. Application use cases depend on these interfaces; the adapter/db
// package provides the concrete implementations.
package port

import (
	"time"

	"github.com/aitrack/server/internal/domain/model"
)

// EditRecordPort is the persistence port for edit records.
type EditRecordPort interface {
	// Save persists a single validated edit record.
	Save(record *model.EditRecord) error
	// CountByTokenKeyAndFilePathSince counts records for rate limiting.
	CountByTokenKeyAndFilePathSince(tokenKey, filePath string, since time.Time) (int64, error)
	// Query returns a page of records plus the total count.
	Query(tokenKey, repoURL string, page, size int) ([]model.EditRecord, int64, error)
	// AggregateByTokenKey aggregates stats grouped by token key.
	AggregateByTokenKey() ([]model.StatsRow, error)
	// AggregateByRepo aggregates stats grouped by repo URL.
	AggregateByRepo() ([]model.StatsRow, error)
	// AggregateByDevice aggregates stats grouped by device ID.
	AggregateByDevice() ([]model.StatsRow, error)
	// AggregateByHostname aggregates stats grouped by hostname.
	AggregateByHostname() ([]model.StatsRow, error)
}

// EditRecordCounter is the narrow port used by the validation domain service
// for rate limiting; EditRecordPort satisfies it.
type EditRecordCounter interface {
	CountByTokenKeyAndFilePathSince(tokenKey, filePath string, since time.Time) (int64, error)
}
